### ISSUE TITLE
Output 240p (instead of 480i) on composite RCA

### ISSUE THREAD

**Issue by starquake** [raspberrypi/firmware#683]
I use my Raspberry Pi to emulate old computers/consoles. A common trick with these systems is the output of 240p. Examples are: ZX Spectrum, Nintendo Entertainment System, Commodore 64. It would be great if I could connect my CRT monitor to it. Right now it only outputs 480i which is interlaced and flickers a lot. 240p removes this flicker and also adds scanlines which make it look more authentic.

More info: http://scanlines.hazard-city.de/

Is it possible to get 240p out of the composite port of the Raspberry PI? Or is there more to it than a change in the firmware?

I think you would make a lot of people even happier than they already are about classic gaming/computing on the Raspberry Pi!

**Comment by popcornmix** [comment]
I'm afraid this has been asked before and it is not supported.
I'm not certain if the hardware supports it or not, but it's certainly never been tested.

There is, a bit (PROG_SCAN) in VEC CONFIG2 register which potentially is useful, but when I asked about it:

> To change from interlaced to progressive, you can't just change a bit in VEC. You also have to turn it off in the HVS and maybe change some parameters in PixelValve, though I'm not sure what. DispmanX probably configures all this, but I forget where.

So I'm afraid there isn't a simple fix for this. It may or may not be possible to make this work, but it's very low priority so I can't promise anything.

**Comment by starquake** [comment]
Thanks a lot for the reply! Wasn't expecting a quick fix.  I'll just keep hoping someday it will be possible.

**Comment by Antiriad** [comment]
I'll keep hoping too. Thanks!

**Comment by starquake** [comment]
I did find a way to output 240p over RGB. More info here:
http://www.retrorgb.com/rpi240pvga.html

In short you can use a Gert VGA and connect it to a Ultimate Scart Adapter.

**Comment by Antiriad** [comment]
@starquake [I'm aware of this](https://forum.recalbox.com/topic/3475/recalbox-sur-tv-crt-en-rgb), thank you. But It would be useful also to output in 240p via Composite.

**Comment by MrBrax** [comment]
Been researching this for the past 3-4 months, went with composite 480i in the end, flickering garbage.

I don't feel very tempted to buy a product (UMSA) that has "Just make sure that your PC outputs VGA lower resolution with 15Khz" in the description..

So I'm really hoping this is gonna work at some point!

**Comment by keilmillerjr** [comment]
+1

**Comment by jimbothegigolo** [comment]
I'm very interested in this as well.
I am hesitant to buy a Raspberry Pi knowing it will only output 480i and the 240p over HDMI is very lackluster.
240p on S-Video, Component or even Composite would be great as well.

Something I've seen on the PSP Go that I found interesting was the way they used the headphone jack.

![37](https://cloud.githubusercontent.com/assets/25341220/22283312/5fdfa316-e296-11e6-873a-e0dd1fc0953a.jpg)

**Comment by MrBrax** [comment]
I've almost achieved it, getting a cheap HDMI->VGA adapter (Powered, important!), and a VGA -> RCA converter i had for my arcade cabinet (it does not produce 240p/15khz however, but it looks a lot sharper), and setting up the resolution to be 240p via hdmi_mode.

Getting an UMSA would be the last step, but i'm not really feeling like buying a €30 gizmo as said earlier. Some countries have stores with cheaper cables but i'm not that lucky to have those.

In a few years, maybe someone will have hacked the firmware on this 👌

**Comment by starquake** [comment]
@jimbothegigolo The Raspberry PI 2 and later also has a headphone jack that also can output composite when using the right cable.

![image](https://cloud.githubusercontent.com/assets/1044544/22283476/c4d8b2f2-e2e2-11e6-9c54-4fa05267da2e.png)

Unfortunately it doesn't output 240p and that's exactly what this request is all about.

**Comment by starquake** [comment]
BTW Arcadeforge is working on this:
http://arcadeforge.net/PiJamma/PI2SCART-Preorder::264.html

And there's also this:
http://www.rgb-pi.com/

**Comment by MrBrax** [comment]
@starquake i've seen that, looks promising! However, that means no GPIO pins left for non-controller use push buttons sadly.

**Comment by jimbothegigolo** [comment]
@starquake I am aware of that, but composite over 480i isn't exactly ideal.

**Comment by Antiriad** [comment]
@MrBrax @keilmillerjr @jimbothegigolo excuse me but the original request is different.

Indeed is yet possible to easily get a 240p by HDMI > VGA > SCART or GPIO (VGA 666) > SCART.

The original request is instead for 240p on **Composite** out (not component) that should allow to get a video mode faithful to some old retro-gaming consoles like NES or Genesis (that had a composite out). A kind of _sdtv_mode=3_  option would be great.

**Comment by MrBrax** [comment]
@Antiriad yes, it would be amazing to do it via composite, but due to the firmware, it doesn't really sound possible. Perhaps in a future hardware revision?

**Comment by Antiriad** [comment]
@popcornmix gave a small glimmer of hope. Even a RP1 can emulate very good NES, Genesis, PC Engine etc. at 60fps in full speed at low resolution (480i) on a CRT TV.
It would be great if,  with some kind of "magic", we could set in config.txt a _sdtv_mode=3_ with, e.g., a 320x224 resolution.
Not 100% accurate but surely much better than 480i.

I want to believe! ;D

**Comment by starquake** [comment]
The console also looks great on 240p. You'll only get something like 80x25 characters, maybe even less but the picture is sharp and stable.

**Comment by starquake** [comment]
@jimbothegigolo 

> I am aware of that, but composite over 480i isn't exactly ideal, if not eye gouging.

I agree. That's exactly what I described when creating this issue.

> There's also the Gert VGA666 if you're willing to settle for VGA. You might be able to convert it to SCART too with a Ultimate SCART Adapter (also from Arcadeforge)

Yep that's what I said in another post and it's pretty much what I'm using.

Don't want to be annoying but you might want to read up on my other posts.

**Comment by keilmillerjr** [comment]
@Antiriad I meant composite. I would have thought that had been an obvious error. I know what I meant to type, but the two words are closely spelled and are both nouns for a type of analogue video transmission.

**Comment by kevinfishburne** [comment]
I've been researching this for a long time now and it blows my mind there is apparently no good solution. Here's the deal: (1) Arcade 15KHz CRTs with RGB/SCART are no longer manufactured and are rare and expensive. (2) NTSC/PAL CRT TV's are also no longer manufactured, but they are cheap and plentiful at thrift stores (Goodwill in the U.S., for example) and they typically have composite and S-Video inputs. (3) Current solutions involve daisy-chaining multiple sketchy adapters; we need something more elegant. (4) HDMI is the most common modern video output format. Conclusion: In the interest of classic gaming preservation, if you do the math, what we need for maximum usefulness to the most number of people is an adapter to take an HDMI video signal and convert it to 240p/288p signal over both composite and S-Video. This will allow anyone with a RPi/PC/laptop/etc. and $10 CRT TV to run classic games exactly as they were intended to look. I say we find a communication hub (Discord/Quip/Facebook/whatever) and start planning a Kickstarter to design and fabricate this active adapter using completely open specifications. Who's with me? I started a Discord server for the idea if anyone wants to take the discussion there: https://discord.gg/dJzGM

**Comment by tekn0x** [comment]
I was able to get pretty close with an off the shelf HDMI to Component cable. This is outputting true 240p. I just have a vertical sync issue. https://www.youtube.com/watch?v=VLBHWZV1dZ4

**Comment by jimbothegigolo** [comment]
@kevinfishburne Your discord invite is expired.

@starquake 
If you use a Pi2SCART you should be able to breakout that signal into S-Video (S-VHS) without much trouble at all, just like there are SCART sockets on some sets that will accept all three composite, S-Video and RGB signals.  
No need to spend hundreds on fancy adapter either.
You might want to buy one of one of those RGB Scart to S-VHS adapters, they go 'round for 2 bucks each and to get composite from it, simply rid the S-Video cable and keep the RCA cables.

**Comment by kevinfishburne** [comment]
@tekn0x - I've seen that video and think it may be an illusion based on the vertical scrolling of the video output. I noticed in certain TG16 games on my RPi on a CRT TV that when the background was scrolling at a speed matching the refresh rate that it created what looked like perfect scan lines. I could be wrong, but if the output were stabilized the scanlines might go away.

@starquake - Apparently Pi2SCART is still taking pre-orders. Is it open-spec? I don't see anything about that. There are HDMI to SCART converters, but no idea if they'll keep vsync and do 240p when daisy chained to an SCART to S-Video adapter. I've been researching this for a while and everything I've found is a mess within a mess to infinity and beyond.

**Comment by starquake** [comment]
@kevinfishburne 

I think Pi2SCART will make an end to at least a part of the mess. The spec of GERT VGA is open. And you there's lot of schematics to find on how to convert VGA to RGBHV. Pi2SCART is just a solution with everything rolled into one.

For now I think the GERT VGA 666 + Ultimate SCART is the best solution.

**Comment by popcornmix** [comment]
I have a test firmware here:
https://drive.google.com/uc?id=0B-6zmEDJwxZEME9lSEFZRmQ5RGs&export=download
 Can you set (in config.txt):
``sdtv_progressive_scan=1``
and report back.
We have a progressive scan bit in the VEC (composite) register set.
We've tested that before and it wasn't sufficient. I've also tried disabling the interlaced signal that gets set. I'm not sure if it does the right thing (my composite TV's quality is a bit too low to tell).

**Comment by MrBrax** [comment]
@popcornmix How would i "install" this?

**Comment by 2skoops** [comment]
popcornmix:
You're amazing!  I just bought an old SOny Trinitron CRT that I'm planning on RGB modding and putting into an arcade cabinet.  I'm still waiting on a Gert's VGA adapter though, so until then I've been playing around with composite 480i... which looks like a total flickering mess!  I happened to check the thread today and saw your new firmware, so I just briefly tested it out.

It's such a huge improvement running this monitor over 240p.  No more flickering, just beautiful stable scanlines.  Everything seemed to be solid for the few minutes I played around with it.  The only thing I noticed was that my picture was noticeably warped and slightly titled near the edges of the screen.  It could just be my TV, since it's opened up right now, but I didn't notice it when I was running 480i.  Could be an issue with the sync over composite... not sure.  I'll have to do a better comparison with 240p vs. 480i on the set tomorrow.  (There was also a high-pitched whine from the speakers, but I'm guess that's just because the TV's guts are exposed right now)

I still plan on moving to 15kHz 240p VGA -> RGB once I get the parts I need, but having a 240p composite hookup is just so convenient and will please a lot of retro gamers!  Thanks again for your hard work!

MrBrax:
Basically you need to copy all of the files into your /boot/ folder.  Probably the easiest way is to power off your Pi, take out the microSD card, and plug it into a card reader on your computer.  Then copy all of the files from the zip over into the boot folder (remember to back up the old versions first).  After that, open /boot/config.txt and add the line popcornmix mentions above:
sdtv_progressive_scan=1
Save, close, eject, put it back into your Pi, and boot it up.

**Comment by Antiriad** [comment]
@popcornmix awesone!
Will it work also on a good old Raspberry Pi 1 B?

**Comment by Henrito** [comment]
@popcornmix 240p composite works with my Raspberry Pi 1 B and 2.

**Comment by Antiriad** [comment]
Thank you @Henrito !!! Can't wait to test it with Lakka!

**Comment by starquake** [comment]
Amazing, will test as soon as possible!

**Comment by danmons** [comment]
Thank you so much, this is a long wanted feature. Will attempt to test on a couple of different boards as soon as possible.

**Comment by MrBrax** [comment]
@2skoops oh i thought i had to flash it. i'll test it the moment i get home from work!!

**Comment by starquake** [comment]
Can anyone post a picture of what it looks like?

**Comment by Henrito** [comment]
@starquake 
![metal_slug_240p_composite_0](https://cloud.githubusercontent.com/assets/15178622/23453645/03c07008-fe1e-11e6-9f0e-d04a31cf6e2c.jpg)
![metal_slug_240p_composite_1](https://cloud.githubusercontent.com/assets/15178622/23453646/03d23e46-fe1e-11e6-85b7-9c8d068f8bb1.jpg)

**Comment by MrBrax** [comment]
Is that really pixel mapped 1:1? Looks sort of blurry.

**Comment by danmons** [comment]
It's still CVBS, so it will still be a little blurry, even if it is progressive scan.  The progressive scan will make moving pictures much better, as you won't get the interlace flicker.

**Comment by starquake** [comment]
SCART RGB will give you better quality. But this will hopefully make the composite out on the Raspberry Pi look like composite on a Famicom for example.

**Comment by MrBrax** [comment]
@danmons yes, but are you sure the scaling/retroarch resolution etc is set correctly to 1:1?

**Comment by starquake** [comment]
@MrBrax what do you mean with 1:1?

Composite is analog. There are no pixels. If we're talking about NTSC there is a resolution of 240 lines vertically. You can use any resolution horizontally. Though you won't be able to see a resolution like 1920. 

@danmons What resolution is output by the Pi? 720x240?

**Comment by keilmillerjr** [comment]
I’m so excited to try this out! Will probably order a new pi3 and the composite breakout cable, so it might be a few weeks for me.

**Comment by MrBrax** [comment]
@starquake but the emulator has options for scaling & aspect ratios, it's not analog until it comes out of the rca connector

**Comment by hmusta** [comment]
@starquake I guess a better way to word the question is:

Let's say you're running at a resolution of 256x224 with a custom hdmi_timings settings. If you set the progressive scan flag for composite video, does it output 240 lines or 224 lines? IIRC, an SNES outputs 224 lines (262 if you include blanking)

**Comment by MrBrax** [comment]
^ retroarch has to be configured to output at that resolution also, with a custom aspect ratio

i can't wait to get home now haha

**Comment by starquake** [comment]
@hmusta @MrBrax  Ah so the question is if the lines are mapped 1:1

I read somewhere it outputs 720x480 in ntsc mode so I guess it will be 720x240 for progressive. Will have to see when I get home. Don't know if hdmi_timings has any effect at all for composite out.

Right now, I use Lakka with RGUI forced to a 4:3 aspect ratio with a 1920x240 resolution without interpolation so both 256x240 and 320x240 look great.

In Lakka I set the system config and the core config to NOT crop the borders so it outputs 240p. It works with most cores. There is a bug with the SNES core though, it doesn't respect the no cropping and scales either 224p to 240p or 239p (not sure). Will have to report a bug about it somewhere someday.

(P.S. we should move this discussion to some forum :-) )

**Comment by MrBrax** [comment]
yep yep

totally depends on the emulators for sure, for nes, they're all widely different on how they crop the picture, and i think either fcemu or nestopia had the least cropping.

well, reddit has a whole bunch of subs/posts about this :)

**Comment by Antiriad** [comment]
@starquake hdmi_timings does not affect composite out, only HDMI or Gert's VGA.

240p via composite will not give a "pixel perfect" video mode for some retro systems, but it's a **GREAT** "quick and easy" solution for a better emulation experience (although not perfect).

For a 1:1 video emulation experience take a look to AdvanceMAME and AdvanceMESS + Gert's VGA 666 + VGA to SCART cable on CRT or Arcade monitor. ;)

**Comment by starquake** [comment]
@MrBrax Just turn off all cropping in system and in the core config. QuickNES and Nestopia will output 240p
Maybe continue discussion here? https://np.reddit.com/r/crtgaming/comments/5wuai2/raspberry_pi_now_support_240p_over_composite/

@Antiriad 
It depends on the width resolution, can't wait to get home

**Comment by netux79** [comment]
What I need to get it working? I get that I need to copy the files to /boot but I'm asking about connecting the RPi to a CRT? What kind of cable I need? Do I need the headphones cable posted above or something else? Thanks

**Comment by hizzlekizzle** [comment]
@netux79 If you have an older RPi with the composite-out jack, you just hook that over RCA cable to your TV. If it's one with the 3.5mm combo jack, you'll need a breakout cable that converts the 3.5mm jack to the standard red/white/yellow RCA plugs.

**Comment by netux79** [comment]
@hizzlekizzle : I have a RPi3, so I think it would be the later option. Thanks hunterk!

**Comment by bparker06** [comment]
>Let's say you're running at a resolution of 256x224 with a custom hdmi_timings setting

@hmusta I don't know how the HDMI hardware in an RPi works specifically, but on typical transmitter chips like Analog Devices they usually have a minimum clock frequency, and it may not like 240p directly. How TVs handle this is described in the EIA/CEA-861 standard, which is to use pixel repetition. So a 320x240 signal over HDMI is actually output over the wire as 2880x240 and then in the AVI InfoFrame packet it specifies what the repetition factor is (9 * 320 = 2880). I think the firmware in an RPi would have to explicitly support that functionality, and I'm not sure it does. Also these "game formats" were not added until revision B of that standard and are completely optional to implement.

**Comment by hmusta** [comment]
@bparker06 since HDMI has a 25MHz lower bound on pixel clock, setting the resolution that way with hdmi_timings won't work if you use the HDMI port and you have to do some pixel multiplication. it does work, however, if you're outputting over VGA with VGA666

**Comment by starquake** [comment]
@popcornmix Works great! Just tested NTSC progressive and I have a stable image.

Lakka thinks it is using a 720x480@60Hz screen which is great because the aspect ratio is correct this way (as opposed to 720x240) I guess every line is skipped?

Did some tests with the 240p suite but I can confirm it does 240p at 60hz. When showing horizontal stripes you see black and white lines, just as is supposed to happen:
![image](https://cloud.githubusercontent.com/assets/1044544/23475442/e34852f6-feb7-11e6-8485-795fa920799e.png)
The striped sprite also show there is no filtering and the 240p works as you would expect:
![image](https://cloud.githubusercontent.com/assets/1044544/23475487/024ab3b0-feb8-11e6-84ac-d14c4bf91171.png)

The flashing sprite also flashes which proves things are running at 60 fps.

For people using Lakka or RetroPie. When scaling the NES width (256) to 720 you can see some slight distortion because 720 is not a multiple of 256 so not every pixel is scaled evenly. You can enable integer scale which resolves this and still matches the original aspect ratio closely. Don't know how that works because 512 is not very close to 720 but the image aspect ratio looks fairly good.

It all looks very much like a Famicom AV

**Comment by jfroco** [comment]
@popcornmix  Thank you! At least.. one of the most wanted features.
Just tested it using a fresh Retropie 4.1 installation and it works great.

I'm playing Castlevania/NES that was impossible to play previously because of the flickering. I will post a video tonight.

**Comment by starquake** [comment]
@jfroco Maybe you already know this but you can use the 240p test suite to see if the 240p output of the emulator core matches up with the 240p output of the pi. If it doesn't you will see thicker black or white lines across the screen.

**Comment by popcornmix** [comment]
I've pushed this change to rpi-update with a slight change.
``sdtv_progressive_scan`` is no more.
You should add 16 onto ``sdtv_mode`` if you want 240p output.
That fits slightly more cleanly with how the firmware handles ``sdtv_mode``.

Examples:
```
sdtv_mode=16    # progressive NTSC
sdtv_mode=0x10  # progressive NTSC (using hex notation)
sdtv_mode=18    # progressive PAL
sdtv_mode=0x12  # progressive PAL (using hex notation)
```

**Comment by MrBrax** [comment]
The screen is just scrolling downwards for me, did i miss some setting? PAL over here.

**Comment by popcornmix** [comment]
@MrBrax are you testing the firmware zip file or the latest rpi-update firmware?
What did you add to config.txt?

**Comment by MrBrax** [comment]
I just did an update but apt-get probably doesn't have the latest one here, do i have to replace certain files from this repo?

**Comment by popcornmix** [comment]
You need to run:
``sudo rpi-update``
but be aware this gets you the bleeding edge firmware.
If you are unsure about doing this then wait for the firmware to reach apt (in perhaps a few weeks).

**Comment by MrBrax** [comment]
ooh rpi-update had to be installed, thought i heard about it somewhere before

okay it's updated the firmware now, and it still spins around
```
sdtv_mode=18
sdtv_aspect=1
```

**Comment by popcornmix** [comment]
Does ``sdtv_mode=2`` also spin?

**Comment by MrBrax** [comment]
nope, it's the one i've been using since the start, and just set it to be sure

**Comment by popcornmix** [comment]
Does ``sdtv_mode=2`` also spin since you updated firmware?
Also report output of ``vcgencmd version``

**Comment by MrBrax** [comment]
uhh `sdtv_mode=16` seems to work, and i'm pretty sure this is a pal tv.. oh well!

emulationstation seems to change the output res when launching a game though, gotta find a fix for that

**Comment by MrBrax** [comment]
Yep, it works perfectly!

Though, i'm still totally confused on how analog TV works, outputting 320x239 with a snes emulator doesn't give the correct aspect ratio, and if i squeeze it vertically, pixels obviously gets smushed.. sheesh

**Comment by popcornmix** [comment]
I've pushed an update if you want to rpi-update again.
It should fix the rolling display with progressive PAL.
Not quite sure why but I had to remove a line of blanking for that.

I've also updated tvservice. Now you can switch at runtime:
```
$ tvservice -c "PAL 4:3 P" ; fbset -depth 8 ; fbset -depth 32; tvservice -s
Powering on SDTV with explicit settings (mode:18 aspect:1)
state 0x80001 [PAL 4:3], 720x576 @ 50.00Hz, progressive
$ tvservice -c "NTSC 4:3" ; fbset -depth 8 ; fbset -depth 32; tvservice -s
Powering on SDTV with explicit settings (mode:0 aspect:1)
state 0x40001 [NTSC 4:3], 720x480 @ 60.00Hz, interlaced
$ tvservice -c "NTSC 4:3 P" ; fbset -depth 8 ; fbset -depth 32; tvservice -s
Powering on SDTV with explicit settings (mode:16 aspect:1)
state 0x40001 [NTSC 4:3], 720x480 @ 60.00Hz, progressive
```

**Comment by jfroco** [comment]
Video of composite output at 240p: https://youtu.be/xEfVIDXnERU

**Comment by keilmillerjr** [comment]
@popcornmix does the update with tv service mean that we could potentially update composite resolutions on the fly?

Having a front end display 480i and then switching to 240p for game play would be amazing.

Thank you so much for this. :1st_place_medal:

**Comment by MrBrax** [comment]
@keilmillerjr i really hope so, it's pretty much impossible to use the interface in 480p downscaled to 240p or 240p native, i'll try it the moment i get home

however, what's more annoying is that applications themselves don't have overscan implementations themselves, so if i want native 320x240, menu items and whatnot gets pushed outside the visible screen area, while the games look great due to scaling options

**Comment by keilmillerjr** [comment]
@MrBrax I am currently using and designing for Attract Mode front end. It can scale to any res, and should look ok assuming the layout was designed with such in mind. My [flavors layout](https://github.com/keilmillerjr/flavors-layout) should work at 320x240, but I didn’t take into consideration overscan. Thanks for bringing that issue up. My neo geo mvs I have very slight crop to fill the viewable area on the crt.

**Comment by popcornmix** [comment]
> @popcornmix does the update with tv service mean that we could potentially update composite resolutions on the fly?

Absolutely. Note the tvservice command will destroy any dispmanx layers.
So if you are using console framebuffer then it needs to be recreated (changing the depth is one way, as used in my example).
If you are using dispmanx or EGL then you should remove dispmanx layers, change tv mode and then add the dispmanx layers again.

In theory you could just spawn a call to tvservice and fbset from your application, but as both are open source, the specific code required could also be embedded directly.

**Comment by MrBrax** [comment]
@keilmillerjr hmm i could try that out, last time i tested it, it was very lackluster, basic, and tedious to configure

**Comment by Antiriad** [comment]
I'm testing the first release of the new firmware (the one on Google Drive) with Lakka and... it's GREAT!
Thank you very very much @popcornmix this is a big improvement on Composite out!

But since Lakka is on OpenElec, there is a way to get the latest firmware release without rpi-update?
Thanks!

**Comment by MrBrax** [comment]
https://youtu.be/Gt1ljJSw57c

After like 3 hours of calibrating with 240pee, i think it's finally done (for NES at least). Seems that NTSC works better than PAL when emulating resolutions.

**Comment by danmons** [comment]
Quick test on an old Model B (not B+) with the Google Drive linked firmware above worked very well for Master System emulation in 240p/NTSC.  Nice solid picture, no eye bleed, even with the limits of CVBS it looked good.  My RetroPie/Jessie packages are way out of date, so I'm updating the entire system and firmware now, and will re-test when updated, and then test an RPi3 later tonight.

**Comment by tekn0x** [comment]
Any chance this could help with the HDMI to component 240p issue as seen in this video? https://www.youtube.com/watch?v=VLBHWZV1dZ4

I know this is for the composite output, but I was just curious.

**Comment by Antiriad** [comment]
So no way to download latest firmware in a zipped archive, to upload it directly in Lakka?
Thanks!

**Comment by popcornmix** [comment]
@tekn0x No, the change applies only to composite output.

@Antiriad That would be a question for lakka. 
You can download latest firmware as a zip [here](https://github.com/Hexxeh/rpi-firmware/archive/master.zip) but I don't know what contents lakka expects.

**Comment by hizzlekizzle** [comment]
@Antiriad 
I'll make sure kivutar is aware of the new firmware and includes it in future images.

**Comment by Antiriad** [comment]
@popcornmix thank you very much. Lakka is built on OpenELEC, so I can't use _rpi-update, apt-get_ or even _sudo_. But with your initial release I copied _bootcode.bin, fixup.dat_ and _start.elf_ on the SD card boot partition and it worked like a charm. Thank you again!

@hizzlekizzle thank you very much, also for your kind feedback on Picodrive core issue!

**Comment by Antiriad** [comment]
I can confirm that the new firmware is perfectly working on the new LAKKA 2.0 RC for Raspberry Pi 1.
I could not be happier ! 👍 

Just curious: why 720x480 and not a "real" 240p? (I mean: there is a real technical limit?)
Anyhow it is better for UIs on CRT TV.

Thanks!

**Comment by jayare5** [comment]
Is it possible to create an image of an SD card with Lakka installed and Composite already enabled so others don't have to mess with it?

**Comment by hizzlekizzle** [comment]
@jayare5 
Once the firmware makes it into LibreElec, Lakka will get it, too. In the meantime, it's pretty painless to get set up:
http://filthypants.blogspot.com/2017/03/raspberry-pi-240p-composite-output.html

**Comment by jayare5** [comment]
@hizzlekizzle 
Thanks! Here's where I'm stuck, in the very beginning:

> you'll need to mount your SD/microSD card on another machine and then find the 'boot' directory (for Lakka, this will be located in the ~500 MB partition that holds the system files 

I have no way to access this. Windows shows "Recovery" partition. No one has ever made mention of this during this process, I'm thinking my SD card is either set up wrong, or there's an obvious tool I'm missing.

**Comment by MrBrax** [comment]
You don't need to put the sd card in a pc and all that, if you have console access on the pi itself, just copy the files to /boot/

**Comment by keilmillerjr** [comment]
I went to radioshack today and got a trrs cable. RPI is weird and yellow and red are swapped. I installed lakka 2.0 rc2 with the help of etcher app for the mac. I used my hdmi to composite cable to set it up. I downloaded the firmware @popcornmix and copied the files mentioned by @hizzlekizzle in filthypants blog. Edited the config file with hdmi_ignore_hotplug=1 and  sdtv_mode=16 as described. Is there any other settings I should be doing? I only have neo geo games, so all are 224 lines.

**Comment by starquake** [comment]
@hizzlekizzle I would like to report an issue about supporting 240p in Lakka. And of course, fixing it. But I need some discussions about what would be the right approach. Do you have any suggestions where to report such an issue? The retroarch repo? The lakkatv/Lakka repo? Or maybe the newer lakkatv/Lakka-LibreELEC repo?

**Comment by hizzlekizzle** [comment]
@starquake 
The issues will likely be in the cores themselves, but you can make the issue at https://github.com/libretro-meta.

**Comment by jayare5** [comment]
Finally I was able to do everything!!
My Raspberry Pi 3 came preinstalled with NooBs, and this was making things different.
After flashing a brand new install, I was able to go and paste the correct files in the correct place easily!
I have, so far, made NES and SNES look nearly identical to their original system, all lines well aligned :D
Also GameBoy (Gambatte) could be made to display similar to a Super GameBoy in size too!
I had to apply a horizontal blur filter (a specific "fast" one that wouldnt slow anything down)

I can't be happier right now!
@starquake Thanks so much for creating this thread and taking your time to report your info!
@popcornmix Thanks so much for providing files!

**Comment by starquake** [comment]
@jayare5 what are you using? Retropie? Lakka?

And what horizontal blur effect are you using? That could fix the artifacts I get with scaling.

**Comment by jayare5** [comment]
Lakka!
It's different for NES and SNES.
To make NES look legit like my real NES, I go to Video settings and choose the "composite" option (Blarrg).
SNES doesnt have this for some reason, but you can apply a shader. I think its in Blurs, and Fast Horizontal.
The name escapes me right now :P sorry, I will look at it later, but Im already having another issue as well...

Lakka seems to be glitchy as hell with these configurations... it often messed up my controllers.
And now it suddenly started mixing shaders and filters between systems even though it wasnt doing that before, so it makes 2 or more filters load at once which makes it slow -.- even if I remove it and tell it to Save core overrides and all the other 2 or 3 save options, it doesnt work.
It's even giving me an "Failed to save to retroarch cfg" thing.

**Comment by njmueller** [comment]
I've got this working pretty well, but I'm running into a bug when calling it programmatically. I've built out a python script that at a certain point executes 'tvservice -c "NTSC 4:3 P"' on the command line. This works perfectly except for when I have imported the pygame module and called pygame.init(). In that scenario, I get a blanked screen, and I can never return to getting picture, even if I call a script that brings me back to HDMI via "tvservice -p"

If I use pygame without that .init() call, it works just fine. Any ideas?

**Comment by popcornmix** [comment]
See: https://github.com/raspberrypi/firmware/issues/683#issuecomment-283634644

**Comment by jayare5** [comment]
@starquake 
Disregard my shader comment, I found a MUCH better one!
I think there can be nothing better than this:

https://github.com/libretro/common-shaders/blob/master/crt/shaders/tvout-tweaks.cg

**Comment by tommik128** [comment]
So how are you going to output custom full width horizontal resolutions like 256, 320 pixels? Can you use custom video modelines? I am afraid that you will not get authentic pixels by using fixed horizontal resolution output like 720. This has to be addressed.

note: Of course vertical resolution is fixed, 240 scanlines at 60 Hz, no discussion about that.

**Comment by danmons** [comment]
> So how are you going to output custom full width horizontal resolutions like 256, 320 pixels?

Remember that for analogue modelines, the horizontal resolution doesn't mean much.  The beam travels across the screen changing colour/brightness as per the timings.  Doubling the hoizontal resolution will make zero difference to what you see.

Line count (vertical resolution) is the value that matters far more.  Aim to make that as accurate as possible, and fit in whatever horizontal resolution you need to keep your pclock and timings happy.

**Comment by tommik128** [comment]
> Doubling the hoizontal resolution will make zero difference to what you see.

Of course. But why would you double the resolution? I was asking about outputting custom horizontal resolution so that you output exact number of unique pixels across the whole visible scanline.

**Comment by starquake** [comment]
@tommik128 Let me summarize what is said in this thread until now: When you enable 240p you get an output resolution of 720x480. 240 lines of those 480 are skipped. (This is useful because you get a nice aspect ratio this way, and you it allows)

About 660 (720-60) of the width is shown, you will lose around 60 pixels because of overscan. This depends on your CRT. The NES has an output of 256x240. I use scaling to show it on the 660x480 on my consumer television. The non-integer scaling is not really visible to me and doesn't bother me. It might be visible on a BVM/PVM.

**Comment by danmons** [comment]
> But why would you double the resolution? 

Because (a) horizontal resolution is somewhat meaningless to a scanline on an analogue display, and (b) your pclock might not be able to get low enough to generate it for very low res systems like the SNES, so doubling the horizontal resolution fixes that, and has zero effect on how it appears coming out of a CRT.

As the scanline draws across the screen, whether it shows one full sized "pixel" or two half sized "pixels" makes zero difference. Doubling the horizontal resolution is an easy way to generate modelines that might not be able to be handled natively. So again, only the vertical resolution (i.e.: the horizontal line count) matters when talking about CRT output.

**Comment by gnomenthusiast** [comment]
Hello,

I have some questions !

I recently got a Pi2Scart for my raspberry Pi, hoping to use with a Sony PVM Monitor.

I have downloaded and installed Lakka, what are the best settings to use in the config.txt file?

Do I need to set up different settings for each emulated system to get the image to fill the CRT screen? 

What is the process for changing settings on Lakka?

Thanks in advance,

GE

**Comment by tommik128** [comment]
> About 660 (720-60) of the width is shown

OK. Very useful information. So that tells me I can display 320 pixels fine, but not 256 pixels.

> The NES has an output of 256x240. I use scaling to show it on the 660x480

Try horizontal scrolling and you will definitely see disproportions unless you use some king of blurring effect - which translates the whole thing into generic VGA PC card with TV out (including Xbox, Wii...). In other words: if the output limits me to 660 horizontal pixels and nothing can be done changing that, I see no other option than RGB interface card like Pi to SCART.

**Comment by MrBrax** [comment]
@tommik128 is correct, if the timings could be set to output 256x240 then it would be 1:1, but it's not, and the width of the picture doesn't scale properly and gets blurry

**Comment by starquake** [comment]
@tommik128 @MrBrax Of course RGB is better. The composite output is pretty close to a Famicom AV though.

I don't see anything weird with vertical lines when scrolling. I do see some issues with the diagonal lines in the mountains of Super Mario. I would think when you scale the 256 pixels over the 660 output and filter it a bit it would look pretty close also. It's not as if the signal of a NES going from a full black pixel to a full white pixel changes like square wave.

**Comment by MrBrax** [comment]
@starquake yeah but it would be optimal if timings could be set manually so it outputs pure 320x240/256x240 (even the output right now doesn't match the x/y/scale of an original nes/snes etc)

**Comment by tomm23** [comment]
alternating white and black pixels produced a slight "rainbow" effect. Is it just me?

**Comment by tomm23** [comment]
Here's what I'm talking about. The real Snes hardware: http://i.imgur.com/W2ELdVk.jpg

Rpi output (stretched horizontally w/ bilinear filtering, btw this is also seen on other emulators like nes and genesis): http://i.imgur.com/7M3RZlM.jpg

Of course the original hardware isn't perfect but it's clearer no doubt. Strange that it has diagonal ripples like this... do other consoles do that?

**Comment by danmons** [comment]
> stretched horizontally w/ bilinear 

What does it look like without bilinear filtering? You shouldn't need that for CRT output, especially for horizontal stretching.

**Comment by tomm23** [comment]
Then I must be doing something wrong because I see a doubling of pixels, hugely distracting .. aren't we trying to scale 320/256/whatever against the final output of 720?

**Comment by tomm23** [comment]
Here are some other pictures to illustrate what I'm seeing.

First, here is the same image as in my previous post, stretched to 320, but without bilinear filtering (you can clearly see the doubling of pixels every 3 or 4 columns): https://s8.postimg.org/tmgwaxhp1/20170315_081921_1.jpg

The next two are not stretched, but at 256, with and without bilinear filtering (they should look the same, as there is no scaling occuring?):
https://s21.postimg.org/p2nmarfjr/20170315_091659_1.jpg
https://s30.postimg.org/otj2dpiup/20170315_091914_1.jpg

Lastly here is a Genesis game (320x244, no filtering), and you can see this ripple effect on the railing.
https://s13.postimg.org/gqn53o0vr/20170315_092635_1.jpg

Any clue as to how to fix this, or is this related to what MrBrax said about the output not being "pure 320x240/256x240", or is it a pixel ratio setting (I've tried them all, to no avail), or is this all due to my tv just being an old piece of junk?

I can minimize the effect to some degree by rendering at 2 or 4 times the resolution...

**Comment by hizzlekizzle** [comment]
CRTs themselves don't have any concept of horizontal resolution, but that stops mattering once we add a fixed output res into the mix, which doesn't allow us to use the console's own pixel clock. That is, if we were able to push out fully custom modelines, we could make the timing match the console's but with a fixed 720x480/240 res, we're stuck with the RPi's hardcoded timing. So, we either stretch the NES/SNES image to fill the available resolution or we have a skinny image with pillarboxing. Genesis/MD, N64 and most PS1 games don't have this issue because their native res is (usually) 320x240 (including overscan) which fits neatly into the RPi output at 2x.

**Comment by MrBrax** [comment]
yes, it's **much** better than before, but getting true console emulation is gonna be... tricky.

-- /boot/config.txt --
disable_overscan=1
framebuffer_width=320
framebuffer_height=240
scaling_kernel=8
sdtv_mode=16
overscan_scale=1
disable_audio_dither=1

-- runcommand setup --
retroarch render res: config

-- nes.cfg --
video_fullscreen_x = "320"
video_fullscreen_y = "240"
custom_viewport_width = "384"
custom_viewport_height = "224"
custom_viewport_x = "-14"
custom_viewport_y = "4"
aspect_ratio_index = "22"
video_crop_overscan = false
video_smooth = false

this produces 1:1 pixels on the rpi, but the pixel clocks/framebuffer res are wrong, so large parts of the image are spilling out on the sides, and the pixels are too large.

**Comment by tomm23** [comment]
Isn't that actually 1.5x1... simply doubling every other column? I still see the same kinda artifacting though, it's merely spread out more and less offensive.

**Comment by kevinfishburne** [comment]
Possibly some useful information here https://retropie.org.uk/forum/topic/8524/component-video-hat/13 (mikechi2) and here https://discord.gg/zqUdBn6 (main chat, also mikechi2). Looks like HDMI timings need to be changed per console / horizontal resolution to properly stretch the width without artifacting (explained by chikechi2 in the Discord chat). Not sure if the HDMI timing issue can be applied equally to the RPi's Composite output or modified real-time by the software/emulator for each console.

**Comment by hizzlekizzle** [comment]
> I have no experience with Wii, but it is said to have the very same limitation of fixed horizontal resolution ouput like over here...

Oh...yeah, I think you're right, actually. It looks like you can change the horizontal res through the [GX video interface](https://github.com/libretro/RetroArch/blob/master/gfx/drivers/gx_gfx.c#L130-L168) but not the actual timing. I'll try to test it some soon.

> Not sure if the HDMI timing issue can be applied equally to the RPi's Composite output or modified real-time by the software/emulator for each console.

I'm pretty sure the Composite output timing is a separate, unrelated thing, which I don't think can be modified at runtime. That is, the 'tvservice' 480i/240p swap isn't really touching the timing, just showing (or not) half of the fields.

**Comment by jfroco** [comment]
I can get almost no artifacts in games with fast 8 dir scrolling like Thunderforce II (Sega Genesis), but with pillarboxing. I think the problem is that I'm using a 4:3 CRT TV, but the RPI is sending 720x480. So, I need 640x480, instead to get rid of the black bars.

@popcornmix  Do you think it is possible to change the output resolution to 640x480? Should it be related to sdtv_aspect=1? Again, thank you for this huge improvement.

(this is the configuration I'm using:

    video_fullscreen_x = "720"
    video_fullscreen_y = "480"
    custom_viewport_width = "597"
    custom_viewport_height = "448"
    aspect_ratio_index = "22"
    video_crop_overscan = false
    video_smooth = false
)

**Comment by MrBrax** [comment]
output res can kinda be tweaked with setting the framebuffer resolution to 640x480 and using the left/right overscan to squeeze the picture in, 720 - 640 = 80/2 = 40px on either side

**Comment by popcornmix** [comment]
@popcornmix Do you think it is possible to change the output resolution to 640x480? Should it be related to sdtv_aspect=1? Again, thank you for this huge improvement.

Have you tried ``framebuffer_aspect``?
See: https://github.com/raspberrypi/firmware/issues/638

**Comment by tomm23** [comment]
framebuffer_aspect does not appear to be doing anything here, and I've tried also changing sdtv_aspect but that only appears to be modifying the console environment and not the actual emulator's output.

**Comment by MrBrax** [comment]
at least if you're using retropie, right before the emulator launches, the aspect/resolution can be set, could be that

**Comment by jfroco** [comment]
Hi @popcornmix , same result as @tomm23 , I tried with fbset and with:

    vcgencmd get_config framebuffer_aspect set 0x40003

It seems that it changes the console output but not inside the emulator.
So far the only solution I have found is to access the service menu in my CRT TV and change the horizontal size to cover the fullscreen... unfortunately the TV I play most doesnt have that option

**Comment by tomm23** [comment]
> using retropie, right before the emulator launches, the aspect/resolution can be set

Has no effect (unless I selecting PAL which produces a scrolling image) and worse yet any setting causes returning to emulation station or console to look like gibberish (flickering desync'ed noisey mess) requiring a reboot.

**Comment by MrBrax** [comment]
are you using the latest latest version? i had that scrolling issue with the google drive zip

**Comment by popcornmix** [comment]
Yes, ``framebuffer_aspect`` will only affect the framebuffer console.

If an emulator uses dispmanx (either directly or through opengles) then that code will need to change to adjust aspect ratio. Effectively the source and destination rectangles passed to ``vc_dispmanx_element_add`` will need to have different aspect ratios (or it could be handled in the opengles code).

This isn't something that can be fixed by the firmware, so you'll need to talk to the emulator devs.

**Comment by tomm23** [comment]
Well, after some research I think the problem I'm seeing (AKA rainbow banding and/or fringing) isn't due to pixel ratios or specific emulators, but ntsc filtering in general. I see it in the console when I set framebuffer to shrink the text to fit more of it onscreen - in which case it kinda looks cleartype-style anti-aliasing but more pronounced.

I read that SNES and NES and some other consoles (inc. later Genesis models?) minimized this **by offsetting the color burst signal** (more info here: http://www.chrismcovell.com/gotRGB/screenshots.html)

Is it possible to achieve this in a firmware adjustment?

In any case, I now feel better knowing I'm not doing something terribly wrong with my setup. I never noticed the rainbow issue on my Genesis playing Sonic. Too bad I just sold that a few months back, otherwise I'd hook it up to see what I've been missing all these years...

**Comment by jayare5** [comment]
That is some great research! I had come across that same site about a year ago too...
I really am not minding the rainbows much over the excitement of already having 240p composite, which already is something many people sadly didnt even care about, so im happy to see people here care about the details of the NTSC artifacts :)
But yeah, I get these rainbows too, they're only noticeable in specific parts of games though.

**Comment by popcornmix** [comment]
I don't believe offsetting the colour burst is possible from the hardware.
The only option I've seen is disabling it (``sdtv_disable_colourburst=1`` in config.txt).
You'll obviously lose colour, but it may be interesting to see if it reduces artefacts.

**Comment by mikechi2** [comment]
I've been playing with scaling on the 240p composite output. It seems like the output is 720 x 480. What I've done (very similar to jfroco) for the SNES is:

Set the emulator render resolution = output resolution. Then in the retroarch.cfg for the system, I set:

video_fullscreen = true
video_smooth = false
video_scale_integer = false
custom_viewport_width = "720"
custom_viewport_height = "448"
aspect_ratio_index = "22"

All smoothing, shaders, etc. should be off.

The idea is to always use integer scaling (224 x 2 for the SNES) in the vertical direction since all consoles must output 240 lines, with the picture occupying some subset. So an integer scaling preserves the ratio between the number of lines in the picture and the total number of output lines.

On the other hand, the number of horizontal pixels is arbitrary and depends on the system. The output should be set equal to the frame buffer width so that it fills the entire width of the screen.

Of course scaling in the horizontal direction is non-integer (720/256 = 2.81) but 720 is plenty of pixels and the distortion is not really noticeable, especially on a CRT.

Results look pretty good:

![img_7126](https://cloud.githubusercontent.com/assets/23427169/24069845/19ca3448-0b6f-11e7-8e52-68a76b54e72c.JPG)

For my component video adaptor, I set a custom output resolution (1600 x 240) and set the viewport width to 1600 which virtually eliminates all horizontal scaling errors. The downside is that the console and GUI looks really funny.

On an unrelated note, I've noticed that the composite video causes the WiFi to become flaky. Also the performance is quite sensitive to the USB cable. On certain cables the angry lighting low voltage warning pops up all the time and the pi slows down. Anyone else observed the same?

**Comment by tommik128** [comment]
> Of course scaling in the horizontal direction is non-integer (720/256 = 2.81) but 720 is plenty of pixels and the distortion is not really noticeable, especially on a CRT.

The thing is: it should not be like that. Using 720 pixels to display 256 pixels is wrong. From my own experience with Xbox emulators: there was no way to have a perfect image of full width 256 pixels mode. Those emulators have test patterns and there was always a distortion if the mapping wasn't integer perfect. If it was integer perfect it was not full width. The only way to mask the problem was to apply a blurr effect. I see the same problem here. Custom horizontal resolution is essential.

**Comment by mikechi2** [comment]
> The thing is: it should not be like that. Using 720 pixels to display 256 pixels is wrong. From my own experience with Xbox emulators: there was no way to have a perfect image of full width 256 pixels mode. Those emulators have test patterns and there was always a distortion if the mapping wasn't integer perfect. If it was integer perfect it was not full width. The only way to mask the problem was to apply a blurr effect. I see the same problem here. Custom horizontal resolution is essential.

Agreed, but for me it's not too bad. For me, scaling 256 into 720 is a lot less objectionable than say 256 into 320.

Another method I tried was to quadruple the frame buffer resolution in the config.txt to 1440 x 960 (viewport at 1440 x 896), which basically made all non integer scaling invisible. However, the GUI was pretty much unusable at that resolution over composite video.

**Comment by hizzlekizzle** [comment]
On a standard-def CRT, I recommend changing the menu in settings > driver from the fancy XMB menu to the retro-styled RGUI menu. It scales down to 240p very nicely. I've been using it on my CRTs for years.

**Comment by ronfar623** [comment]
@popcornmix Hey, I just wanted to thank you sincerely for your work on this, and if you take donations (bitcoin or otherwise) I, and I'm sure a few others, would love to buy you a beer, just let us know where.  I bought a RPi3 as a direct result of this firmware, then another once I discovered how useful they are, and am now considering a third to build into an arcade stick.  I'm sure I'm not alone in feeling that this has the potential to be the holy grail of retro console emulation.

Sorry if this is off-topic, but I also wanted to ask if anyone knows of the easiest/cheapest method of getting S-video out of the Pi (or if it's even possible)?  240p over the composite output is already leaps and bounds better than the old 480i, but there are a ton of televisions and PVMs with S-video, and I know from experience that the quality difference can be remarkable.  Thanks!

**Comment by mikechi2** [comment]
@ronfar623 I'm working on an a hat that does component video and S-video:

![image](https://cloud.githubusercontent.com/assets/23427169/24086048/0a8fb2bc-0cc5-11e7-9ba0-df96adee7ca5.png)

Here's a video of the prototype doing 240p over component:
https://www.youtube.com/watch?v=1m8FpY-mk8A

**Comment by jayare5** [comment]
@ronfar623 I'm one of those who would donate to this as thanks :)
Just like you, I bought a Raspberry Pi 3 just because of this, and I'm also considering buying a few more and maybe gift some friends!
My only sadness is that people who own standard def TVs have become hard to find, meaning only the few purists like us will be able to enjoy this. Besides, HD TVs can be laggy... SD TVs are the best for retro games!
I've spent a lot setting up my roms and settings for all the systems I like to emulate.
I wish I (or someone) could develop a core that acts as a (fake?) Super GameBoy that can run easily on such low hardware, unlike BSNES, so that we can have the borders, sounds, menus and processing speed / sound pitch of the Super GameBoy!

**Comment by danmons** [comment]
@mikechi2 - your component hat looks amazing.  Very glad you're working on this.  I'll be buying a few for sure!

**Comment by mikoyv3** [comment]
@popcornmix I appreciate your work. question, is this included in the newly release retropie 4.2? if not, will it still work? 

@mikechi2 I am eagerly awaiting your product, please take my money!

**Comment by hizzlekizzle** [comment]
@mikoyv3 
[As mentioned earlier in the thread](https://github.com/raspberrypi/firmware/issues/683#issuecomment-283478710), if you're using raspbian, you can run:
sudo rpi-update
and it will fetch the latest firmware.

**Comment by mikechi2** [comment]
Thanks for your kind words everyone! :) Here's the latest:

![img_7150](https://cloud.githubusercontent.com/assets/23427169/24135934/c74d252e-0dc9-11e7-8234-a64f791424ae.JPG)

https://www.youtube.com/watch?v=6abdzDPUuzU

The latest PCB came back and most of it works quite nicely. Still have a few minor bugs to shoot down. Notably, this one seems to cause the RPi to throw the angry yellow low voltage warning, which doesn't make sense to me... Also need to play around with the output voltage levels to ensure compatibility with all TVs.

S-Video works really nicely. On my 13in PVM, I can't really tell the difference between S-Video and RGB. On my 27in Trinitron, I think component's colors are a bit punchier. Both are very sharp.

Composite is quite a bit worse mainly because the 3.57MHz chroma subcarrier isn't synchronized with the RPi's pixel clock. The problem is compounded by the fact that the RPi's pixel clock can't generate an exact 15.734 kHz horizontal scan rate, which messes up the TV's comb filter. In my experience, using the 240p firmware and the RPi's built in composite output is a bit better.

Will keep you guys posted!

PS - Whoever is running this thread. If I'm hijacking the topic, let me know..

**Comment by mikechi2** [comment]
@hizzlekizzle Yeah, I actually got 1600 x 240 to work. It worked well for eliminating horizontal scaling errors but the GUI and everything else was completely unusable. This is asking a lot but an idea solution might be a) have the GUI's display account for the distortion and provide a normal aspect ratio output or b) have some sort of on the fly display output switching. In the meantime, I'm plenty happy just playing with a letterboxed screen or tolerating some fuzziness in the horizontal axis.

**Comment by MrBrax** [comment]
I think it's time for a new RPi 240p distro for retro games ;)

**Comment by starquake** [comment]
@MrBrax and everyone else. May I suggest continuing discussion on how we can improve retroarch/lakka/retropie here: https://github.com/libretro/libretro-meta/issues/73

**Comment by evanrinehart** [comment]
@mikechi2 will you be releasing the designs for that board or selling units? Either way, I want one for my "new" trinitron badly!

**Comment by mikechi2** [comment]
@evanrinehart @mikoyv3 @danmons thanks guys, check out here for more info: [www.retrotink.com](www.retrotink.com) :)

**Comment by mikoyv3** [comment]
@mikechi2 I'm so happy you are making this available, paid and sent my order for the RetroTINK-C. 

oh no, just realized I may want the **RetroTINK-RGB** instead. How different is this version? Would I be able to notice a difference?

**RetroTINK-RGB** 
Reference quality RGB output
Texas Instruments triple video DAC

**Comment by evanrinehart** [comment]
I was going to buy one of each, but since the form doesn't let you combine them (and thus save on shipping) I ordered two -S's. Nice work. Fingers crossed!

**Comment by peasquared** [comment]
Hi all, I recently updated using sudo rpi-update and now I'm unable to access any RetroPie options via the GUI. If I choose any of them, (Wifi Settings, ES Themes, etc) I get this "Unknown Platform" error for a second before kicking me back out. Does anyone have any ideas what this error means?
![img_4604](https://cloud.githubusercontent.com/assets/26757789/24437998/6ac5fbc4-13f9-11e7-82aa-142ff0fefe23.JPG)

**Comment by starquake** [comment]
@peasquared This is a github issue related to a 240p firmware. I think you will have more luck if you go to the RetroPie support forums: https://retropie.org.uk/forum/

**Comment by peasquared** [comment]
@starquake Thanks, I'll do that! I just figured I'd try here, since it happened after doing the sudo rpi-update, trying to get 240, but now that I think about it, doing the update doesn't only affect 240p firmware.

**Comment by danieljg** [comment]
@tekn0x Your adapter could work correctly if you set the refresh rate to 29.94 Hz.

Please let me know if this worked for you, I'd love to get one of those for myself if it did.

**Comment by tekn0x** [comment]
@danieljg Any suggestions on how to set it to 29.94 Hz?

**Comment by danieljg** [comment]
@tekn0x Try running the following commands;

tvservice --ntsc -e "CEA 8"
fbset -depth 8 && fbset -depth 16

(the second line may have to be introduced 'blind', if there is no output after the first. it may be best to place this in a startup script)

Hope this helps.

**Comment by tekn0x** [comment]
@danieljg Unfortunately I no longer have access to a CRT television to test with.

**Comment by danieljg** [comment]
That's a shame, I'm pretty sure this would fix it. Please let us know if you get a chance to test in the future.

**Comment by starquake** [comment]
I got my 240p output as I wished. Closing this.

**Comment by starquake** [comment]
Big thanks to @popcornmix !

**Comment by tomm23** [comment]
Was there some recent development, or are you simply OK with there being no adjustable composite timings?

**Comment by starquake** [comment]
I got what I asked for. So I closed it. And I thought that for the other things discussed in this issue, separate issues should be created.

**Comment by hizzlekizzle** [comment]
I'm not sure adjustable composite timing is even in the realm of possibilities.

**Comment by sdubrick** [comment]
> I've been playing with scaling on the 240p composite output. It seems like the output is 720 x 480. What I've done (very similar to jfroco) for the SNES is:
> 
> Set the emulator render resolution = output resolution. Then in the retroarch.cfg for the system, I set:

@mikechi2 I've been trying to set up my raspi built-in composite output, but facing horrible scaling issues.  I'm not sure if I'm following these directions correctly - to clarify, by setting the emulator render resolution = output resolution, do you mean by setting it to the raspi's output resolution (720x480)?  So, does this mean setting video_fullscreen_x = "720" and video_fullscreen_y = "480" in retroarch.cfg? Thanks!

**Comment by starquake** [comment]
@sdubrick Here's my config. I have no noticeable scaling issues: https://gist.github.com/starquake/4b53a9da62f6ad4681079b6b25e8c23b

**Comment by sdubrick** [comment]
@starquake Thanks! That fixed all scaling issues for the most part.  The only thing I haven't tried setting yet from your configs was the tvout-tweaks shader since I didn't have that installed.  On the snes emulation I tried there's quite a bit more vertical overscan on the top/bottom (around 20 pixels total), and none on the sides - is that to be expected?

**Comment by starquake** [comment]
@sdubrick, would like to see that. Can you take a picture?

**Comment by sdubrick** [comment]
@starquake I think I figured it out - I'm not sure why I needed to, but after I tweaked some configuration settings from the ones you provided and I'm now getting a better picture.  Here's what I did (screenshots from SNES 240p test suite)

This was originally how it looked after applying the configuration - not horrible, but you can see it seemed slightly more vertically stretched as you can see from the overscan
![img_20170420_204034](https://cloud.githubusercontent.com/assets/5289502/25295838/f826fba8-26aa-11e7-8197-aec5457c603e.jpg)
![img_20170419_222214](https://cloud.githubusercontent.com/assets/5289502/25295887/36498626-26ab-11e7-8f3c-ac7f6bc9fd85.jpg)

I changed the aspect_ratio_index from config to custom and that fixed the aspect ratio, but it seemed like all sides were still a bit too overscanned.  
![img_20170420_212955](https://cloud.githubusercontent.com/assets/5289502/25295964/95ace446-26ab-11e7-9ea2-b686e6da65f4.jpg)

In config.txt I commented out the disable_overscan=1 line and added some overscan_top/ overscan_bottom/overscan_left/overscan_right values until I was happy with it:
![img_20170420_223424](https://cloud.githubusercontent.com/assets/5289502/25296115/471f0da8-26ac-11e7-9cb3-f1bb8dfedcab.jpg)

Note I haven't tried any other system emulators on it, but SNES is my main concern and it seems to look ok

The only minor issue is that there is still some inaccuracy in scaling as you can see in the horizontal line test below.  But I don't notice it when playing, so I'm not too bothered by it. Does it happen to you?
![img_20170420_204356](https://cloud.githubusercontent.com/assets/5289502/25296239/e3173262-26ac-11e7-9c36-707e2b5cecf5.jpg)

**Comment by tomm23** [comment]
You should be able to get perfect horizontal lines... Remember SNES games output 224, so if the emulator is outputting 480 you'll wanna use 448 for the custom window height. There will usually be overscan (parts of the game you cannot see) and that is normal

**Comment by manekinekodesu** [comment]
**Big thanks** to @popcornmix for this awesome feature. Bought my rpi3 specifically for this.
But I have a **question**, I'm not sure to ask here or to open another issue: 
Can rpi3 be  configured to output **PAL60** signal over composite? (It's different from Brasilian PAL-M, which is not supported by my PAL crt tv).

**Comment by edmaul69** [comment]
@popcornmix will this mess up using hdmi? i want to be able to switch back and forth

**Comment by hizzlekizzle** [comment]
No, HDMI is unaffected.

**Comment by oguzcantasci** [comment]
First of all, I want to thank everybody who has somehow contributed to this topic. I have read the whole topic for 3 times, but I still could not resolve the issues I have. I want to ask some questions to clear up my mind and achieve what I want.

1-) First question is about my CRT tv. I cannot understand whether it is Pal or NTSC. No manuals or info on the web. At the back it is written 50/60 Hz. tvservice -s shows PAL 720x576 progressive as I set hdtv_mode=18. It is probably PAL but I wanted to specify this confusion I have.

2)What I am really trying to achieve is the perfect scanlines. , but whatever I tried mentioned above did not work for me. I can get perfect scanlines with RGUI custom aspect ratio but that is almost the half of the screen gone.

3) Let's say my tv is PAL. What are the resolutions I should set in config.txt and in retroarch.cfg in order to get almost full screen perfect scanlines. I want to get to know the logic behind resolution setting better.

If you need more info from me I am 7/24 here as I am really into getting this to work.

Thank you in advance.

**Comment by MrBrax** [comment]
@oguzcantasci some tv's support both pal and ntsc, but it's most likely pal

i ran it with 320x240 framebuffer resolution but it doesn't look very good, the overscan kills it

**Comment by hizzlekizzle** [comment]
If it's a PAL-only TV and you're running with PAL timings and resolution, you'll probably want to run PAL games, as well.

**Comment by oguzcantasci** [comment]
Okay now, here is the deal.
I uninstalled retripie now, installed lakka.
Set sd_tv to 16 for NTSC mode.
And with custom aspect ratio something like 642 to 477 I managed to get perfect scanlines as it was real nes hardware. But That probably means tv is compatable with both ntsc and pal.
But I want to get the same perfect scanlines with PAL to as it has higher resolution and better colors. Do you have any idea how can I get the same fof PAL output ?

**Comment by MrBrax** [comment]
>Do you have any idea how can I get the same fof PAL output ?

I tried for four months with no avail, i'll be amazed if someone finds out. It's gotta view the 240pee stripes properly and show everything in a correct aspect ratio (not 320x240 1:1)

**Comment by manekinekodesu** [comment]
Hello @oguzcantasci ! I've spent a few months tweaking paremeters in the retroarch.cfg and conf.txt, and finally got perfect image on my pal crt tv. I've also have a dendy(pal famicom clone) and sega megadrive2 and the rpi composite output is indentical, video quality even slightly better than on the originals. I will upload my retroarch.cfg and conf.txt a bit later if you are interested. You also need to run pal versions of the games obviously.

**Comment by manekinekodesu** [comment]
Here are the files:
https://github.com/manekinekodesu/lakka-PAL50-config
@oguzcantasci let me know if it works for you or you have any questions ;-)

**Comment by oguzcantasci** [comment]
Thank you for your files. I copied your files to the respective directories. Ran the snes 240pee PAL test. I cannot get perfect scanlines. It is like double. Two black scanlines and then two white, on and on. What is wrong with my setup I am about to go crazy, Please someone help me.

**Comment by oguzcantasci** [comment]
I am getting a felling that my TV might be actually NTSC after all. Why would not it work with the same config files as yours :/

**Comment by manekinekodesu** [comment]
@oguzcantasci its really strange that you get double scanlines. I've tried my setup on different crts, one of them being multi sysyem pal/ntsc. Maybe there are other config files that I've missed.
It woud be helpful if you could take a high res photo of your crt's screen(with scanlines being visible). Then i could provide you with some constructive feedback, I hope. 
Also try using Nestopia for testing, as the snes is a bit more tricky to setup correctly.

**Comment by oguzcantasci** [comment]
@manekinekodesu hey man thank you for your help. I will get photos as soon as I can. Actually I think retropie is kind of faulty I guess. Because LAkka straight out of the box for me just with adjusting custom aspect ratio. Really weird tho :(

**Comment by oguzcantasci** [comment]
@manekinekodesu OKAY now it all comes down to this. I am going ntsc. I am starting to explain.
- My main example is Castlevania. 
-When using Lakka and its default settings, I set the aspect ratio manually to 642x448, and some xy custom positions and it is identical to a real NES. I got this from a youtube video here : https://www.youtube.com/watch?v=p8_eKmmN5rA&t=520s ( settings at 21:00 minute mark)
 But when get to the retropie with sdtv_mode=16 and load it up it is like this.
![whatsapp image 2017-08-11 at 02 51 51](https://user-images.githubusercontent.com/30830211/29196976-c2d3b700-7e40-11e7-9c3e-3b7814afcd6d.jpeg)
In lakka, when  in the video it is compared to NES , ''P'' of Player , ''S'' of Score and ''E'' of enemy is visible. And I am able to get this with Lakka too, but I want to use retropie.
Well you can say, adjust the x position then ! Well when I try, I see that the remaining left side is not there at all like this
![whatsapp image 2017-08-11 at 02 51 51 1](https://user-images.githubusercontent.com/30830211/29197014-0b3d440c-7e41-11e7-9e8f-013a32887b27.jpeg)

What can be the main reason or difference for this ? What would you recommend ? Framebuffer or overscan in config ? Do they effect the emulators ? Thank you in advance

**Comment by tomm23** [comment]
Paste your config.txt settings here? There are overscan positioning values in there that so effect retroarch as well as emulation station (when it's properly set up, the edges of the default ES theme will be unreadable)

**Comment by tomm23** [comment]
> rpi composite output is indentical
I thought horizontal resolution was impossible to match because of differences in hardware timing... in other words a diagonal line would appear to have little stair steps in it, unless you use smoothing filter which would make it look a bit fuzzy... I'd love to proven otherwise though

**Comment by kevinfishburne** [comment]
FYI my experimentation/calibration was using Lakka, not RetroPie, so YMMV if using my settings in the "The One You've Been Waiting For" video. Horizontal resolution is impossible to match, although "ultrawide" modes can be used although not for the 240 firmware mod which is fixed at 720x480. There is now a horizontal-only filtering shader which should alleviate the pain of using a width of 720: http://eightvirtues.com/lakka/hat/Shaders/

**Comment by dankcushions** [comment]
@oguzcantasci 
your castlevania cropping issue is nothing to do with CRT settings - it's an emulator settings issue. see https://github.com/retropie/retropie-setup/wiki/Nintendo-Entertainment-System#how-to-disable-crop-overscan-in-lr-fceumm - lakka and retropie may have different default settings or different NES emulators

i think this thread is straying into general support and it would be better to migrate to https://www.reddit.com/r/crtgaming/ or https://retropie.org.uk/forum/ (etc). github isn't really the place for this, IMO.

**Comment by oguzcantasci** [comment]
I FINALLY GOT IT WORKING. 
I will write here what I have done to inform people who might have the same problem someday. 
I set it to sdtv_mode=16 for ntsc. Don't set sdtv_aspect to anything it messes it up.
While in a nes rom, I select+x'ed for retroarch gui. I went for the quick menu, then options, and there I disabled any kind of overscan.
Also in video settings disable the crop overscan.
After these, I managed to get perfect scanlines and an identical output like NES by adjusting custom aspect ratio. However it is 640 something x 480, so I am not sure it is perfectly like NES as it has a different resolution. But 240pee test suite shows perfect scanlines both for vertical and horizontal.

**Comment by kevinfishburne** [comment]
When viewing games on a CRT TV, which naturally overscans, you want to disable in the emulator settings as many of the overscan cropping settings as possible so it shows as much of the game as possible, then adjust the custom viewport settings appropriate and save as a core override in the quick menu.

**Comment by Regulations** [comment]
@oguzcantasci 
I have edited config.txt to disable_overscan=1 (uncommented)
I´m trying without overscan by disable everything in retroarch and adjusting custom aspect ratio. Everything is fine in emulators but how do I set custom aspect ratio in Emulationstation? 
My menu is too far to the left.

**Comment by tomm23** [comment]
Emulationstation uses themes defined in XML. They can be modified (which is what did) but there may be some already set to work nicely on your screen.

**Comment by tomm23** [comment]
Nobody can get perfect scanlines, but if the slow scrolling image test does not "ripple" then it means you aren't skipping or duplicating pixels. You'll probably want to use bilinear filtering or shaders to smooth them out though. It won't match the console hardware perfectly but some here seem satisfied enough with it

**Comment by danieljg** [comment]
@tomm23 what do you mean nobody can get perfect scanlines?

If you have a real 240p output, then there will always be real scanlines on your screen irrespective of the image you're sending. It might be that there isn't 1:1 pixel mapping, so some lines could be shown doubled, but the scanlines won't be affected by this.

**Comment by tomm23** [comment]
Ah, I thought it was clear he had been using the phrase "perfect scanlines" as a catch-all to include pixel mapping as it compares to the real console hardware, so I was just trying to drive home that it cannot be 100% the same but you can get sorta close.

**Comment by Regulations** [comment]
@tomm23 Could you point me to some good Emulationstation themes? Which one are you using?
Do you have any example to look at where the "ripple" shows so I know what to look for?

@danieljg How do you check if you get real 240p output in retropie?

What happens if you have disable_overscan=1 and still adjust by overscan_left/right/top/bottom, does that remove the purpose?

I really want to get this as good as possible.
I also set up framebuffer_width=320 framebuffer_height=240 to get retropie-setup readable on my CRT.

**Comment by oguzcantasci** [comment]
I got all my problems sorted. My tv is ntsc. Other than emulation and perfect scanlines, I set emulation station to run at a spesific resolution to fit the screen, and I edited the Comicbook 4:3 theme to fit my crt perfectly. I will upload my config and all the other extra information needed in a day's time.

**Comment by danieljg** [comment]
@Regulations it should be as simple as looking at the crt.

If the image flickers like it does when you play a dvd on that same screen, it's interlaced, if the image is dead stable, like you'd get on a NES, then you have a progressive image.

Look it up on youtube, it's not hard at all to tell.

**Comment by jayare5** [comment]
Is there a way to make the 240p behave a little differently?
I just finally realized something is wrong with this but wasn't sure what. This could potentially increase the quality of the image.
Normally, game consoles use an interlacing technique even though they output 240p, each line updates differently (maybe it alternates between the odd and even lines?) and it makes some NTSC artifacts less noticeable. These are all static at least in the 240p firmware I have and it makes games with white text look like they have red dots around them. Of course, I use a filter which makes it look REALLY close to the original systems, but maybe somewhere in the firmware there's an option to make it behave even more like original videogame consoles? We might need popcornmix opinion on this :P unless this has already been discussed?

**Comment by hizzlekizzle** [comment]
There are some things that consoles did to minimize artifacts such as rainbowing in Sega Genesis/MD games, but I don't believe it's possible to reproduce those strategies in the firmware because they depend on the exact timing of the console in question.

**Comment by oguzcantasci** [comment]
@jayare5 which filter are you using, sir?

**Comment by jayare5** [comment]
@oguzcantasci 
Depends on the system, I have shader tvout-tweaks as the all-around reliable! There are others that work better in specific cases but terrible in other cases. TVout-tweaks also for some reason makes the black not as pure black. Makes me a little hesitant to recommend.
Wish I could develop my own shader :(
All you really need is a slight horizontal pixel softening. That's it. Im assuming it must be difficult to develop because almost all the "Blur horizontal" ones can look really bad on this.

@hizzlekizzle 
Ahhhh I see! I mean, this is super amazing as it is already!
Maybe the only improvement at this point is really for a new mini computer to be made that has more flexible 240p, but considering how long it took to get the RPi to do it, I won't hold my breath xD
The videogame console I've seen with the cleanest and best 240p output is the PlayStation 2 (I have slim, not sure about phat one)
But to be honest, I DO love the NTSC artifacts in some cases, like for emulating NES games, because the NES did show quite a noticeable amount of it too, so in the end I still like the Pi EXACTLY as it is now :D

Again, thanks so much @popcornmix !!!!!!

**Comment by Clive75** [comment]
Hey guys, I need some help with my Pi3. Trying to output in 240p through the RCA to my NTSC CRT. I updated to the latest firmware, added sdtv_mode=16 to the config file, but the TV is still outputting interlaced. Tried mode 0,1,2,18 and even tried to switch to PAL and somehow its not working, when I typed tvservice -s and it tells me its outputting in 720x480 60hz interlaced despite the changes in config file. I have no clue what is wrong.

**Comment by starquake** [comment]
@Clive75 You say you added a line, but there already should be one IIRC. Does it have 2 lines now? If so, remove one.

**Comment by Clive75** [comment]
I dont think there was one.
Here's my config file, just in case

https://gist.github.com/Clive75/ba425fff946893603a0fe511341de3e6

**Comment by ultraover** [comment]
@Clive75 Did you update manually or used the rpi-update?

**Comment by starquake** [comment]
@ultraover You did see it was fixed, right?

**Comment by ultraover** [comment]
@starquake yo, as a matter of fact, i was reading your post at libretro #73

i will be honest here, I asked that question because i am really confused about implementations and updates involving the rpi-update. Take issue #811 for instance, i have no ideia if I can get changes discussed there using the rpi-update instead of downloading the files popcornmix linked to.

Now, being more specific: i am running batocera linux 5.9 and trying to make 240p over composite. The question is: will rpi-update give me what it takes to enable "new" sdtv_modes?

I am new to the community and to the whole pi3 scene, so bear with me for a while :)

**Comment by Clive75** [comment]
@ultraover I used rpi-update. Lost a few hours because of some spaces but lesson learned.

**Comment by ultraover** [comment]
@Clive75 thanks!

**Comment by HellTazer** [comment]
So i can output 240p on a sony wega crt just using Clive75 configs and the the 3.5mm jack of the Pi3? over my composite AV? no need for adaptors? thanks to all :)

**Comment by GerryHolt** [comment]
Is there a way to change the resolution while the machine is running or to set up a script to reboot the machine in a different resolution based on what's launched?

**Comment by sdubrick** [comment]
@HellTazer Yes
@GerryHolt Yes, see [this commen](https://github.com/raspberrypi/firmware/issues/683#issuecomment-283519135)t for the commands to switch resolution during runtime

**Comment by jayare5** [comment]
Does anyone here use ASUS' Tinker Board? It has very good specs but doesn't output video for CRTs through the Audio jack. I was wondering, how will we do it moving forward with possible better / future boards that only have HDMI?
![tinker board](https://user-images.githubusercontent.com/26237704/33237857-2bbe4658-d24d-11e7-9ad7-43fd9577b546.jpg)

**Comment by hizzlekizzle** [comment]
@jayare5 We'll just have to use the costlier GPIO-based output hats.

**Comment by jayare5** [comment]
@hizzlekizzle Huh! So that's possible for 240p?
Also even though the Pi3 B is already capable of emulating almost every game console that originally used 240p, I wonder if a more powerful computer would have even better performance like reducing input lag, which in that case it would be worth the cost I think :D

**Comment by jayare5** [comment]
@hizzlekizzle Huh! So that's possible for 240p?
Also the cost would be worth it if the games that can already be emulated on the Pi3 B could take advantage of the more powerful hardware! (Like reducing input lag even further.)

**Comment by KoolKiller** [comment]
Thanks for all the effort, especially to popcornmix. 

Is there a way to enable the firmware fix on a PAL system? I get it to work in black and white/NTSC, but when I set sdtv_mode=2, the pcture begins to scroll continuously from top to bottom (otherwise it looks fine). I'm on a CRT monitor.

**Comment by manekinekodesu** [comment]
Hey @KoolKiller, maybe check out this issue [Support for PAL60 in config.txt/sdtv_mode](https://github.com/raspberrypi/firmware/issues/811)

sdtv_mode=2 will get you standard interlaced pal 720x576p output.
sdtv_mode=18 will get you progressive pal 720x288p output.
sdtv_mode=34 is for pal60, but, sadly,  it doesn't work

**Comment by KoolKiller** [comment]
@manekinekodesu, thanks for the tip!

Sadly, sdtv_mode=18 doesn't fix it in connection with sdtv_progressive_scan=1. I assume that it works fine on NTSC. Has anybody successfully tested it on a PAL CRT monitor? I'm on a Commodore 1084S, if that helps. 
I would really like to figure this out. My only other output device is a HD projector. 8-bit looks fine, but anything above tends to strain my eyes.

**Comment by manekinekodesu** [comment]
I'm using PAL CRT TV. And pal50 progressive mode works just fine (sdtv_mode=18 or sdtv_mode=0x12).
I'm not sure, if you need sdtv_progressive_scan=1 line.
@KoolKiller, You can compare your config file with mine here(it's a bit messy, sorry):
https://github.com/manekinekodesu/lakka-PAL50-config/blob/master/flash/config.txt

**Comment by Rayanfer32** [comment]
The composite pin on my pi zero is not responding ..... was working fine before  ....and now its like pin is dummy ... i tried all possible ways to make it work ... the problem is now i want to change the default TV out pin config ... any idea

**Comment by jayare5** [comment]
@starquake @alessioscand @MrBrax @jimbothegigolo @kevinfishburne @tekn0x @KoolKiller @danieljg @danmons @hizzlekizzle 

Hey, are you guys still interested in this? Check this out, we've made a Lakka branch specifically for the Composite output of Raspberry Pi, with updated RetroArch + cores!

https://lakka.tv/articles/2024/05/02/rpi-composite/

**Comment by ibex-are-goats** [comment]
@jayare5 Thank you for the composite output release of Lakka!!!!

**Comment by jayare5** [comment]
We've made important updates for PAL TVs! If anyone was having trouble getting a progressive signal, you should redownload from that same article and please read the changes made to the "PAL users" section for instructions.

**Comment by sharmashivanand** [comment]
Is this still possible with Pi 4 in 2024? Which version of Raspberry OS should I be using?
